fix: correlation remover docs

taharallouche opened the pull request:
## Description
Hi @TamaraAtanasoska @hildeweerts , I’ve updated the mathematical formulation in the doc to better match the optimization problem that the `CorrelationRemover` class is actually solving. WDYT ? :pray: 

## Tests
<!--- Select all that apply by putting an x between the brackets: [x] -->
- [x] no new tests required
- [ ] new tests added
- [ ] existing tests adjusted

## Documentation
<!--- Select all that apply. -->
- [ ] no documentation changes needed
- [x] user guide added or updated
- [ ] API docs added or updated
- [ ] example notebook added or updated

## Screenshots [OUTDATED]
![image](https://github.com/user-attachments/assets/0164bcba-373a-473a-89b2-b3b14987f50c)

hildeweerts commented:
Many thanks for thinking this through and making a PR @taharallouche. This formulation follows the code more closely and is hopefully easier to understand.

To further enhance clarity, it might be useful to briefly explain some of the variables before you introduce the mathematical formulation (e.g. $W$ the vector of coefficients of the LR model, $\bar{S}$ average of $S$, and $Z^*$ the transformed non-sensitive features) and possibly add a sentence in between $min_W$ and $\texttt{let}  Z^*$ to make clear these are two separate steps.

@MiroDudik @koaning were involved in the original PR of this method and may have thoughts!

taharallouche commented:
@hildeweerts I hope these changes address your suggestions. I had to go more into details to accurately describe what [numpy's `lstsq`](https://numpy.org/doc/stable/reference/generated/numpy.linalg.lstsq.html) is doing under the hood (solving the least squares column-wise).

I'll be happy to make any other changes (including rolling back to a higher-level description :sweat_smile: ).

MiroDudik commented on docs/user_guide/mitigation/preprocessing.rst:
```diff
@@ -20,23 +20,43 @@ Sensitive features can be correlated with non-sensitive features in the dataset.
+:code:`alpha` parameter.
+
+In mathematical terms, assume we have the original dataset :math:`X`, which
+contains a set of **sensitive attributes** denoted by :math:`S` and a set of
+**non-sensitive attributes** denoted by :math:`Z`. The goal is to remove
+correlations between the sensitive attributes and the non-sensitive attributes.
```
```suggestion
contains a set of **sensitive features** denoted by :math:`S` and a set of
**non-sensitive features** denoted by :math:`Z`. The goal is to remove
correlations between the sensitive and the non-sensitive features.
```

MiroDudik commented on docs/user_guide/mitigation/preprocessing.rst:
```diff
@@ -20,23 +20,43 @@ Sensitive features can be correlated with non-sensitive features in the dataset.
+
+Let :math:`m_s` and :math:`m_{ns}` denote the number of sensitive and non-sensitive
+features, respectively.
+Let :math:`\bar{S}` represent the mean of the sensitive attributes, *i.e.*,
```
Let's use "sensitive features" and "non-sensitive features" rather than attributes.

The feature mean is a vector, not a matrix. It might be better to write it as `\bar{\mathbf{s}}`, and it would be probably more standard if it was a column vector (although its dimension is the number of columns of the matrix S).

For the vectors, I'd go with the boldface, e.g., `\mathbf{s}` (but the same would apply to all other vectors in this PR). For consistency, I think that we should do the same for the matrices and write `\mathbf{S}` and `\mathbf{Z}`. What do you think?

MiroDudik commented on docs/user_guide/mitigation/preprocessing.rst:
```diff
@@ -20,23 +20,43 @@ Sensitive features can be correlated with non-sensitive features in the dataset.
+:math:`\bar{S} = (\bar{s}_1, \dots, \bar{s}_{m_s})`, where
+:math:`\bar{s}_j` is the mean of the :math:`j\text{-th}` sensitive feature.
+
+For each non-sensitive feature :math:`z_j` in :math:`Z`,
```
Maybe we can say something like `\mathbf{z}_j\in\mathbb{R}^n` where `j=1,\dotsc,m_{ns}` to drive home the dimension of `\mathbf{z}_j`.

MiroDudik commented on docs/user_guide/mitigation/preprocessing.rst:
```diff
@@ -20,23 +20,43 @@ Sensitive features can be correlated with non-sensitive features in the dataset.
+Let :math:`m_s` and :math:`m_{ns}` denote the number of sensitive and non-sensitive
+features, respectively.
+Let :math:`\bar{S}` represent the mean of the sensitive attributes, *i.e.*,
+:math:`\bar{S} = (\bar{s}_1, \dots, \bar{s}_{m_s})`, where
```
If we go with the column vector for `\bar{\mathbf{s}}`, then we may want to write it as `[\bar{s}_,1, \dots, \bar{s}_{m_s}]^\top`.

MiroDudik commented on docs/user_guide/mitigation/preprocessing.rst:
```diff
@@ -20,23 +20,43 @@ Sensitive features can be correlated with non-sensitive features in the dataset.
-The amount of correlation that is removed can be controlled using the
-:code:`alpha` parameter. This is described as follows:
+
+    \min _{w} \| z_j - (S - \bar{S}) w \|_2^2
```
In numpy, this works thanks to broadcasting, but here we have a dimension mismatch, so we would need to write `\mathbf{S}-\mathbf{1}\bar{\mathbf{s}}^\top` where `\mathbf{1}` is the all-ones vector in `\mathbb{R}^n`.

MiroDudik reviewed (commented):
Some clarity suggestions to incorporate.

taharallouche commented:
Hi @MiroDudik thanks for the review ! I believe I addressed all your suggestions (I just took the liberty to use `\mathbf{1}_n` to emphasize that it is a vector of `n` ones, all the rest follows the review) :ok_hand: 

PS: the flaky test is breaking the CI pipeline, again :pensive:

TamaraAtanasoska commented:
> Hi @MiroDudik thanks for the review ! I believe I addressed all your suggestions (I just took the liberty to use `\mathbf{1}_n` to emphasize that it is a vector of `n` ones, all the rest follows the review) 👌
> 
> PS: the flaky test is breaking the CI pipeline, again 😔

hey yes unfortunately those problems are known :( For a while it was better as the server response improved on the OpenML side and I implemented caching, but I might need to check how persistent it is as it gets deleted after a while. One thing I can suggest is to push an empty commit with `git commit --allow-empty -m "Trigger Build"` and then a push. There were times when the server response is so bad it took me several tries, I couldn't get to the caching as it just won't download. It is much better than before atm, but if it becomes unmanageable again we can think of doing something more.

MiroDudik commented on fairlearn/preprocessing/_correlation_remover.py:
```diff
@@ -32,28 +32,51 @@ class CorrelationRemover(BaseEstimator, TransformerMixin):
-    values. To describe that mathematically, let's assume in the original dataset :math:`X`
-    we've got a set of sensitive attributes :math:`S` and a set of non-sensitive attributes
-    :math:`Z`. Mathematically this method will be solving the following problem.
+    values. To describe that mathematically, let's assume in the original dataset
+    :math:`\mathbf{X}` we've got a set of sensitive attributes :math:`\mathbf{S}` and a set of
+    non-sensitive attributes :math:`\mathbf{Z}`.
```
```suggestion
    features. To describe that mathematically, let's assume in the original dataset
    :math:`\mathbf{X}` we've got a set of sensitive features :math:`\mathbf{S}` and a set of
    non-sensitive features :math:`\mathbf{Z}`.
```

MiroDudik replied:
```suggestion
    features.
```

MiroDudik commented on fairlearn/preprocessing/_correlation_remover.py:
```diff
@@ -32,28 +32,51 @@ class CorrelationRemover(BaseEstimator, TransformerMixin):
-    values. To describe that mathematically, let's assume in the original dataset :math:`X`
-    we've got a set of sensitive attributes :math:`S` and a set of non-sensitive attributes
-    :math:`Z`. Mathematically this method will be solving the following problem.
+    values. To describe that mathematically, let's assume in the original dataset
+    :math:`\mathbf{X}` we've got a set of sensitive attributes :math:`\mathbf{S}` and a set of
+    non-sensitive attributes :math:`\mathbf{Z}`.
+
+    In mathematical terms, assume we have the original dataset :math:`\mathbf{X}`, which
```
```suggestion
    features. In mathematical terms, assume we have the original dataset :math:`\mathbf{X}`, which
```

MiroDudik reviewed (commented):
This looks good. I just caught some leftover text that should have been deleted in the doc (see the suggested edit).

MiroDudik reviewed (approved):
This looks good from my side--I just noted one leftover text that should have been deleted in the doc (see the suggested change). Once that is incorporated this is good to go.

taharallouche commented:
> This looks good from my side--I just noted one leftover text that should have been deleted in the doc (see the suggested change). Once that is incorporated this is good to go.

Thanks for spotting that :pray: It's fixed now :ok_hand:
